### ISSUE TITLE
delete test class ltcl_dangerous in ZCL_ABAPGIT_OBJECTS 

### ISSUE THREAD

**Issue by larshp** [abapGit/abapGit#3953]
https://github.com/abapGit/CI should be used instead

**Comment by FreHu** [comment]
there is also a checkbox in the settings to enable these tests

**Comment by larshp** [comment]
yea, that can also be removed after the usage in https://github.com/abapGit/abapGit/pull/3711 is removed

**Comment by pokrakam** [comment]
> yea, that can also be removed after the usage in #3711 is removed

Not if we want to invoke CI or similar functions from the IDE. I like things simple and am used to pressing ctrl-shift-F10 to test as I go along.

CI absolutely has it's benefits, but for TDD running a report is not an option and having Eclipse call the API is outside my skillset. That's also the topic of #3776 

IMO a great solution would be to get CI automated in a way similar to ADT's background syntax checking, or have a dedicated shortcut key.

**Comment by flaiker** [comment]
> > yea, that can also be removed after the usage in #3711 is removed
> 
> Not if we want to invoke CI or similar functions from the IDE. I like things simple and am used to pressing ctrl-shift-F10 to test as I go along.
> 
> CI absolutely has it's benefits, but for TDD running a report is not an option and having Eclipse call the API is outside my skillset. That's also the topic of #3776
> 
> IMO a great solution would be to get CI automated in a way similar to ADT's background syntax checking, or have a dedicated shortcut key.

I don't think abapGit/CI or integration tests in general are suitable for TDD. The last run published on ci.abapgit.org took 20 minutes...

**Comment by pokrakam** [comment]
My de/serialization test runs in half a second :) 

![image](https://user-images.githubusercontent.com/25539927/94460184-4517ed80-01b0-11eb-9a56-8e96fb25f30e.png)
 (NPL in VMWare, 2 year old mid-spec MacBook Pro)

Integration vs Unit testing is not entirely clear-cut. The problem is that creating an SAP object is not feasible to mock, so real de/serialization is the only option for testing an object class.  
The way I see it a single object de/serialisation constitutes a unit test, multiple objects is more integration. I can quickly check if I've broken anything in my class or I can use CI to verify the whole app still works.

**Comment by flaiker** [comment]
On 7.53 there's something similar where you can use `@testing` to relate a test class to a different object. And if you press Ctrl+Shift+F10 on the object it should execute the test. Just fyi, cannot use it here of course.

https://help.sap.com/doc/abapdocu_754_index_htm/7.54/en-US/index.htm?file=abendoccomment.htm Chapter "Test relations"

Having integration tests inside the test include is fine by me, I just don't execute them TDD-style but rather have them run nightly and ideally isolated in a test client or even test system that can be restored to the state before the tests. Otherwise you may run into the problem that failed tests leave behind a state that prevents further test runs from being executed. And of course not mocking the external dependencies results in lock situation on shared systems or influences not accounted for in the test setup (date, time, weather).

**Comment by pokrakam** [comment]
Yes I read about test relations, the unit test book also covers them, handy feature when we finally get our hands on it. 

I completely agree with all your comments, and in most cases I try to avoid going to the DB, especially writes. 
But sometimes the only alternatives are not to mock or not to test. abapGit Objects are a little special because we can't mock SAP's black box because that is also part of what we're testing. I had similar issues trying to unit test Gateway DPC classes, /IWBEP/CL_MGW_REQUEST is just too complex and unpredictable to mock. 

There's also a difference between a unit test that creates a Sales Order with delivery and invoice, and a TDD-friendly create/delete operation. For my scenario it took the better part of an hour to write the test in question, and it saved me far more time than that. Easy to write, easy to understand, quick to run - makes it mostly a unit test in my book. 

Re. Risk, this is a point we can discuss at length. I've had abapGit corrupt objects several times, e.g. leaving a report in a state where it's visible in SE80 but can't be edited or deleted because it doesn't exist. So the risk is there from the word go. Is there much difference between someone deliberately switching on "Dangerous tests" and running all the abapGit unit tests and someone installing and running CI? In both cases it's a deliberate act by someone knowingly taking a risk. 

Mind you, I didn't know about CI until a month or so ago, so this could be hilighted in the contribution guidelines. When I did hear about it, I looked at it, thought it was a pretty cool setup, but saw talk about how long it takes to run and that it was last run in 2019 for that reason, so that gave a clear signal to write my own UT if I want to have something TDD-friendly.

But as I also said in the other issue, I'm keen on trying to get to a best of both worlds scenario where we can use the same basis for TDD and CI. Happy to spend time on that after I'm done with my current bits.
